pkg/trace/agent/agent_test.go

```diff
@@ -1551,3 +1553,346 @@ func TestSampleWithPriorityNone(t *testing.T) {
 	assert.True(t, keep) // Score Sampler should keep the trace.
 	assert.EqualValues(t, numEvents, 0)
 }
+
+// TestSpanSampling verifies that an incoming trace chunk that contains spans
+// with "span sampling" tags results in those spans being sent to the trace
+// writer in a sampled (kept) chunk. In other words, if a tracer marks spans as
+// "span sampled," then the trace agent will not discard those spans.
+//
+// There are multiple cases to consider:
+//
+//  1. When the chunk would have been kept anyway (sampling priority > 0). In
+//     this case, the tracer should not have added the span sampling tag.
+//     Regardless, verify that the chunk is kept as expected.
+//  2. When the chunk is dropped by the user, i.e. PriorityUserDrop sampling
+//     priority. In this case, other samplers do not run at all, but the span
+//     sampler still runs. The resulting trace chunk, if any, contains only
+//     the spans that were tagged for span sampling, and the sampling priority
+//     of the resulting chunk overall is PriorityUserKeep.
+//  2a. Same as (2), except that only the local root span specifically is tagged
+//       for span sampling. Verify that only the local root span is kept.
+//  2b. Same as (2), except that only a non-local-root span specifically is
+//      tagged for span sampling. Verify that only one span is kept.
+//  3. When the chunk is dropped due to an agent-provided sample rate, i.e. with
+//     PriorityAutoDrop priority. In this case, other samplers will run. Only if the
+//     resulting decision is to drop the chunk, expect that the span sampler
+//     will run and yield the same result as in case (2).
+//  4. When the chunk is dropped due to an agent-provided sample rate, i.e. with
+//     PriorityAutoDrop priority, but then the error sampler decides the keep
+//     the chunk anyway. In this case, expect that the span sampler will _not_
+//     run, and so the resulting chunk will contain all of its spans, even if
+//     only some of them are tagged for span sampling.
+func TestSpanSampling(t *testing.T) {
+	// common returns boilerplate used by all test cases below.
```

Comment:
There's a way to write this test that's a bit more "Go"-y. If you want me to show you in a call, I'm happy to do so. Otherwise, here's the high level suggestion I would make is to use a [table test](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) for this. It would look something like this:

```
    tests := []struct {
        name string
        payload &pb.TracerPayload
        checks func(*testing.T, []*pb.TraceChunk)
    }{
        {
            name: "case 1: would have been kept anyway",
            payload: ...
            checks := func(t *testing.T, chunks []*pb.TraceChunk) {
                .....
             }
        {
.....
         },
    } 

    for _, tc := range tests {
        t.Run(tc.name, func(t *testing.T) {
    		cfg := config.New()
    		cfg.Endpoints[0].APIKey = "test"
    		// Disable the rare sampler. Otherwise, it would consider every first
    		// priority==0 chunk as rare. Instead, we use the error sampler to
    		// cover the case where non-span samplers decide to keep a priority==0
    		// chunk.
    		cfg.RareSamplerDisabled = true
    		ctx, cancel := context.WithCancel(context.Background())
    		defer cancel()
    		traceAgent := NewAgent(ctx, cfg)
    		traceAgent.Process(&api.Payload{
			    // The payload might get modified in-place, so first deep copy it so
			    // that we have the original for comparison later.
			    TracerPayload: proto.Clone(tc.payload).(*pb.TracerPayload),
			    // a nil Source would trigger a panic
			    Source: traceAgent.Receiver.Stats.GetTagStats(info.Tags{}),
    		})
    		assert.Len(t, traceAgent.TraceWriter.In, 1)
    		sampledChunks := <-traceAgent.TraceWriter.In
		    chunks := sampledChunks.TracerPayload.Chunks
    		tc.checks(t, chunks)
    	}
    }
```

where the function in `checks` would be any test-specific checks

Reply:
Okeydokey.